Telemetric support

tomasg2012 opened the pull request:
Adds support for SubmitTestMetricReport, with persistent reports and collection creation

Clarified certain variable names

Moved code outside of for loop to guarantee one execution.

Fixed missing parameter in do_DELETE

tomasg2012 committed:
added telemetric support from preliminary code

Added line to store metrics report to @odata.id

Name cleanup, moved code to avoid repeat execution

mraineri commented on redfishMockupServer.py:
```diff
@@ -430,6 +432,92 @@ def do_POST(self):
+                            self.send_response(404)
+                        else:
+                            # Check if all of the parameters are given
+                            if ('Name' not in data_received) or ('Value' not in data_received):
```
Should be "MetricReportName" and "MetricReportValues". Looking further at the TelemetryService, there seems to be a discrepancy in data types too that we need to look at in the forum...

mraineri replied:
With the current schema definition, we could just generate a single value in a metric report; this may be okay. So, the POST would look like:
```
{
    "MetricReportName": "TestReport",
    "MetricReportValues": "6"
}
```

and the generated report would look like:
```
{
    "@odata.context": "/redfish/v1/$metadata#MetricReport.MetricReport",
    "@odata.type": "#MetricReport.v1_0_0.MetricReport",
    "@odata.id": "/redfish/v1/TelemetryService/MetricReports/TestReport",
    "Id": "TestReport",      <--- Comes from "MetricReportName" input
    "Name": "TestReport",    <--- Comes from "MetricReportName" input
    "ReportSequence": "<SomeSequenceNumberIncremented>",
    "MetricReportDefinition": {
        "@odata.id": "/redfish/v1/TelemetryService/MetricReportDefinitions/TestReport"
    },
    "MetricValues": [
        "Timestamp": "<current time>",
        "MetricValue": "6"    <--- Comes from "MetricReportValues" input
    ]
}
```

@jcleung5549 was this the original intent, or is the action supposed to take in the entire array of values specified by "MetricValues"?

mraineri commented:
Will need some follow up based on reviewing the existing schema

mraineri commented on redfishMockupServer.py:
```diff
@@ -430,6 +432,92 @@ def do_POST(self):
+                                    item = {}
+                                    item['MemberID'] = tup[0]
+                                    item['MetricValue'] = tup[1]
+                                    item['TimeStamp'] = tup[2]
```
Should be lower case S (Timestamp)

mraineri commented on redfishMockupServer.py:
```diff
@@ -430,6 +432,92 @@ def do_POST(self):
+                                        'Name']}
+                                for tup in data_received['Value']:
+                                    item = {}
+                                    item['MemberID'] = tup[0]
```
Should be "MetricId"

tomasg2012 committed:
Simplified/extracted repeated code, corrected bugs

Removed stdout flushes, unused code/parameter (translate_path, testetagflag, load)

Reduced repetition of path/grab code (construct_path, try_to_sleep, cache...)

Moved POST actions to their own functions (event, telemetry, collections)

Fixed infinite loop in Member creation

Added threaded handling of POST commands (grequests library)

New requirement (grequests)

Fixed values for telemetric action

Added requirements for grequests

Fixed mising threading target + params

Replace prints with automatically flushed logger

mraineri commented:
With the new schema updates, the expectation is the server can take a payload in this format:

```
{
    "MetricReportName": "TestReport",
    "GeneratedMetricReportValues": [
        {
            "MetricId": "SomeId",
            "MetricValue": "20",
            "Timestamp": "2019-02-05T18:00:00Z",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Power#/SomeProperty",
            "MetricDefinition": { "@odata.id": "/redfish/v1/MetricDefURI" }
        },
        {
            "MetricId": "SomeId",
            "MetricValue": "25",
            "Timestamp": "2019-02-05T18:00:05Z",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Thermal#/SomeProperty",
            "MetricDefinition": { "@odata.id": "/redfish/v1/MetricDefURI2" }
        }
    ]
}
```

This would result in the following type of Metric Report:

```
{
    "Id": "TestReport",
    "Name": "Metric Report",
    "Timestamp": "TimestampOfWhenThisReportIsAssembled",
    "MetricValues": [
        {
            "MetricId": "SomeId",
            "MetricValue": "20",
            "Timestamp": "2019-02-05T18:00:00Z",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Power#/SomeProperty",
            "MetricDefinition": { "@odata.id": "/redfish/v1/MetricDefURI" }
        },
        {
            "MetricId": "SomeId",
            "MetricValue": "25",
            "Timestamp": "2019-02-05T18:00:05Z",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Thermal#/SomeProperty",
            "MetricDefinition": { "@odata.id": "/redfish/v1/MetricDefURI2" }
        }
    ]
}
```

tomasg2012 committed:
Error checks around PATCH/POST Payloads

tomasg2012 committed:
Updated telemetric model to latest

tomasg2012 commented:
Added exception wrapping for POST and PATCH
```
POST: Headers: Host: localhost:8000
Connection: keep-alive
Content-Length: 19
Authorization: Basic cm9vdDpjYWx2aW4=
Postman-Token: 4b7bb5db-d8b4-c498-e8be-e8c144343485
Cache-Control: no-cache
Origin: chrome-extension://fhbjgbiflinjbdggehcddcbncdddomop
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36
Content-Type: application/json
Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9


Decoding JSON has failed, sending 405
127.0.0.1 - - [07/Feb/2019 15:02:49] "POST /redfish/v1/Chassis HTTP/1.1" 405 -
```

Added more recent 2018.3 definition of SubmitTestMetricReport

resulting payload:

```
{
    "@odata.context": "/redfish/v1/$metadata#MetricReport.MetricReport",
    "@odata.id": "/redfish/v1/TelemetryService/MetricReports/TestReport",
    "@odata.type": "#MetricReport.v1_0_0.MetricReport",
    "Id": "TestReport",
    "MetricReportDefinition": {
        "@odata.id": "/redfish/v1/TelemetryService/MetricReportDefinitions/TestReport"
    },
    "MetricValues": [
        {
            "MetricDefinition": {
                "@odata.id": "/redfish/v1/MetricDefURI"
            },
            "MetricId": "SomeId",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Power#/SomeProperty",
            "MetricValue": "20",
            "Timestamp": "2019-02-05T18:00:00Z"
        },
        {
            "MetricDefinition": {
                "@odata.id": "/redfish/v1/MetricDefURI2"
            },
            "MetricId": "SomeId",
            "MetricProperty": "/redfish/v1/Chassis/SomeChassis/Thermal#/SomeProperty",
            "MetricValue": "25",
            "Timestamp": "2019-02-05T18:00:05Z"
        }
    ],
    "Name": "TestReport",
    "Timestamp": "2019-02-07T14:16:56-43"
}
```

tomasg2012 committed:
Changed 405 to 400

mraineri commented:
Approved 2/7
